File: backend/src/main/java/com/eduproctor/tut/service/CategoryService.java
package com.eduproctor.tut.service;

import com.eduproctor.tut.entity.Category;

import java.util.List;

public interface CategoryService {
    public Category createCategory(Category category);

    public List<Category> getAllCategories();

    public Category getCategoryById(Long id);

    public void deleteCategory(Long id);

    public Category updateCategory(Long id, Category category);
}

File: frontend1/src/app/services/category.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Category } from '../models/category.model';


@Injectable({
  providedIn: 'root'
})
export class CategoryService {

  private baseUrl = 'http://localhost:8082/api/categories';

  constructor(private http: HttpClient) { }

  addCategory(category: Category): Observable<Category> {
    return this.http.post<Category>(this.baseUrl, category);
  }

  getAllCategories(): Observable<Category[]> {
    return this.http.get<Category[]>(`${this.baseUrl}`);
  }

  getCategoryById(id: number): Observable<Category> {
    return this.http.get<Category>(`${this.baseUrl}/${id}`);
  }

  public deleteCategory(catid: number): Observable<any> {
    return this.http.delete(`${this.baseUrl}/${catid}`);
  }

  public updateCategory(id: number, category: Category): Observable<Category> {
    return this.http.put<Category>(`${this.baseUrl}/${id}`, category);
  }
}
